frontend/crd/src/app/milestones-page/milestones/milestone.service.ts

```diff
@@ -1,26 +1,46 @@
 import { Injectable } from '@angular/core';
-import {HttpClient} from "@angular/common/http";
-import {Milestone, MilestoneJSON} from "../../../domain/Milestone";
-import {map, Observable} from "rxjs";
+import { HttpClient } from "@angular/common/http";
+import { Milestone, MilestoneJSON } from "../../../domain/Milestone";
+import { concatMap, finalize, map, Observable, of, ReplaySubject } from "rxjs";
 import { Endpoints, constructBackendRequest } from 'src/app/util/http-helper';
 
 @Injectable({
   providedIn: 'root'
 })
 export class MilestoneService {
 
+  private milestoneCache$ = new ReplaySubject<any>();
+  private hasBeenRequested = false;
+
   constructor(
     private http: HttpClient,
   ) {
   }
 
-  // Get all data items
-  getMilestones(): Observable<Milestone[]> {
-    return this.http.get<Milestone[]>(constructBackendRequest(Endpoints.MILESTONES))
-      .pipe(map((data: any) => {
-        return data.map((milestoneData: MilestoneJSON) => {
-          return new Milestone(milestoneData)
-        })
-      }))
+  getMilestones(forceRefresh?: boolean): any {
+    // return last value (i.e. cache) from ReplaySubject or add data to it
+    if (!this.hasBeenRequested || forceRefresh) {
+      this.hasBeenRequested = true;
+      const milestoneData = this.http.get<Milestone[]>(constructBackendRequest(Endpoints.MILESTONES))
+        .pipe(map((data: any) => {
```

Comment:
make this a .subscribe instead and then just update the cache here. you don't need any of the stuff bellow

Reply:
the http method is async (i think????) so any time i tried to do that the code just moved on without waiting to receive any data and the cache is empty